packages/lit-html/src/directives/async-replace.ts

```diff
@@ -10,69 +10,127 @@ import {AsyncDirective} from '../async-directive.js';
 
 type Mapper<T> = (v: T, index?: number) => unknown;
 
-class AsyncReplaceDirective extends AsyncDirective {
-  private _value?: AsyncIterable<unknown>;
-  private _reconnectResolver?: () => void;
-  private _reconnectPromise?: Promise<void>;
+// When a directive is connected and actively awaiting a result from a promise,
+// the directive will be stored in `promiseToDirectiveMap`. The resolver will
+// get a reference back to the directive to perform the commit via the WeakMap
+// rather than closing over the directive instance, which would hold the
+// directive until the promise resolved.
+const promiseToDirectiveMap: WeakMap<Promise<unknown>, AsyncReplaceDirective> =
+  new WeakMap();
+
+// When a directive is disconnected and an awaited promise resolves, the
+// result will be stored in `promiseToResultMap` such that it can be retrieved
+// if/when it reconnects
+const promiseToResultMap: WeakMap<Promise<unknown>, IteratorResult<unknown>> =
+  new WeakMap();
+
+export class AsyncReplaceDirective extends AsyncDirective {
+  private __iterable!: AsyncIterable<unknown>;
+  private __iterator!: AsyncIterator<unknown>;
+  protected index = 0;
+  private __nextPromise?: Promise<IteratorResult<unknown>>;
+  private __mapper?: Mapper<unknown>;
 
   // @ts-expect-error value not used, but we want a nice parameter for docs
   // eslint-disable-next-line @typescript-eslint/no-unused-vars
   render<T>(value: AsyncIterable<T>, _mapper?: Mapper<T>) {
     return noChange;
   }
 
-  update(_part: ChildPart, [value, mapper]: DirectiveParameters<this>) {
+  update(_part: ChildPart, [iterable, mapper]: DirectiveParameters<this>) {
     // If we've already set up this particular iterable, we don't need
     // to do anything.
-    if (value === this._value) {
+    if (iterable === this.__iterable) {
       return;
     }
-    this._value = value;
-    this.__iterate(mapper);
+    // If we're starting a new promise, stop tracking the last one, since its
+    // result is no longer relevant
+    if (this.__nextPromise !== undefined) {
+      promiseToDirectiveMap.delete(this.__nextPromise);
+    }
+    this.__mapper = mapper;
+    this.__iterable = iterable;
+    this.index = 0;
+    // Note, we use the iterator protocol manually (rather than for/await/of) so
```

Comment:
This is pretty tricky. It'd be really nice to have a `DisconnectablePromise` directive-helper to make this more generally available.

Reply:
With a helper, you can actually continue using `for await`. The issue is that we're inside a concise method, which strongly holds onto the `this` value. If we defer to something like an `asyncForEach`, we've broken the inherent memory leak. We'll still need to indirect to get the `directve` value, but that's much simpler than writing manual iteration:

```js
async function asyncForEach(iterable, cb) {
  for await (const v of iterable) {
    if (await cb(v) === false) break;
  }
}
```

Adding in `=== false` allows us to break the iteration, if needed for any reason. And the `await cb(v)` would allow us to await a pause action, much like the current code:

<details>

<summary>AsyncPause</summary>

```js
class AsyncPause {
  _pause = null;
  _resolve = null;

  get() {
    return this._pause;
  }

  pause() {
    if (this._pause) return;
    this._pause = new Promise(resolve => {
      this._resolve = resolve;
    });
  }
  
  resume() {
    this._resolve?.();
    this._pause = null;
    this._resolve = null;
  }
}
```

</details>

```js
class AsyncDirective {
  _pauser = new AsyncPause();

  update(iterable) {
    const pauser = this._pauser;
    asyncForEach(iterable, async (value) => {
      if (pauser.get()) await pauser.get();

      // indirect to get directive here.
      if (directive._value !== iterable) return false;
    });
  }
}